File: python/fields.py
"""
Shared field-definition engine for FFX / FFX2 save data.

A `Field` describes one named value inside the save body (numeric fields
are little-endian; `kind='bit'` fields pack a single flag into one byte).
All offsets are relative to the start of the *core save body* - i.e. after
any platform header has already been stripped (see save_data.py, which
adds `header_len` back on for you).
"""

from dataclasses import dataclass
from typing import Optional
import struct

_STRUCT = {
    "u8": ("<B", 1),
    "u16": ("<H", 2),
    "u32": ("<I", 4),
    "i32": ("<i", 4),
    "f32": ("<f", 4),
}


@dataclass(frozen=True)
class Field:
    name: str
    offset: int
    kind: str  # 'u8' | 'u16' | 'u32' | 'i32' | 'f32' | 'bit'
    bit: Optional[int] = None  # only for kind='bit': which bit of the byte at `offset`
    note: str = ""


def bit_field(name, base_offset, bit_index, note=""):
    """Mirrors FFXED's BitFieldAccessor: a "bit index" can run past 7 and
    spill into later bytes (byte = base + bit_index // 8, bit = bit_index % 8)."""
    return Field(name, base_offset + bit_index // 8, "bit", bit_index % 8, note)


def read_field(data, field: Field, header_len: int = 0):
    off = header_len + field.offset
    if field.kind == "bit":
        return bool(data[off] & (1 << field.bit))
    fmt, _size = _STRUCT[field.kind]
    return struct.unpack_from(fmt, data, off)[0]


def write_field(data: bytearray, field: Field, value, header_len: int = 0):
    off = header_len + field.offset
    if field.kind == "bit":
        if value:
            data[off] |= 1 << field.bit
        else:
            data[off] &= ~(1 << field.bit) & 0xFF
        return
    fmt, _size = _STRUCT[field.kind]
    struct.pack_into(fmt, data, off, value)

File: typescript/src/lib/fields.ts
// Shared field-definition engine for FFX / FFX2 save data - a straight
// port of the Python project's fields.py. A FieldDef describes one named
// value inside the save body (numeric fields are little-endian; kind
// "bit" fields pack a single flag into one byte). All offsets are
// relative to the start of the core save body - i.e. after any platform
// header has already been stripped (see saveData.ts, which adds
// headerLen back on for you).

export type FieldKind = "u8" | "u16" | "u32" | "i32" | "f32" | "bit";

export interface FieldDef {
  name: string;
  offset: number;
  kind: FieldKind;
  bit?: number; // only for kind "bit": which bit of the byte at `offset`
  note?: string;
}

function dataViewFor(data: Uint8Array): DataView {
  return new DataView(data.buffer, data.byteOffset, data.byteLength);
}

export function readField(data: Uint8Array, field: FieldDef, headerLen = 0): number | boolean {
  const off = headerLen + field.offset;
  if (field.kind === "bit") {
    return (data[off] & (1 << (field.bit as number))) !== 0;
  }
  const view = dataViewFor(data);
  switch (field.kind) {
    case "u8":
      return view.getUint8(off);
    case "u16":
      return view.getUint16(off, true);
    case "u32":
      return view.getUint32(off, true);
    case "i32":
      return view.getInt32(off, true);
    case "f32":
      return view.getFloat32(off, true);
  }
}

export function writeField(data: Uint8Array, field: FieldDef, value: number | boolean, headerLen = 0): void {
  const off = headerLen + field.offset;
  if (field.kind === "bit") {
    if (value) {
      data[off] |= 1 << (field.bit as number);
    } else {
      data[off] &= ~(1 << (field.bit as number)) & 0xff;
    }
    return;
  }
  const view = dataViewFor(data);
  const num = value as number;
  switch (field.kind) {
    case "u8":
      view.setUint8(off, num);
      break;
    case "u16":
      view.setUint16(off, num, true);
      break;
    case "u32":
      view.setUint32(off, num, true);
      break;
    case "i32":
      view.setInt32(off, num, true);
      break;
    case "f32":
      view.setFloat32(off, num, true);
      break;
  }
}

export function coerceValue(raw: string, kind: FieldKind): number | boolean {
  if (kind === "bit") {
    const s = raw.trim().toLowerCase();
    return s === "1" || s === "true" || s === "yes" || s === "on";
  }
  if (kind === "f32") return parseFloat(raw);
  return parseInt(raw, 10);
}
